OrbitBase/Tracing.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "OrbitBase/Tracing.h"
+
+#include <absl/base/internal/sysinfo.h>
+
+#include <memory>
+#include <vector>
+
+#include "OrbitBase/Logging.h"
+
+using orbit::tracing::Scope;
+using orbit::tracing::ScopeType;
+using orbit::tracing::TimerCallback;
+
+thread_local std::vector<Scope> scopes;
+
+TimerCallback* g_callback;
+
+// Dummy implementation. This depends on another PR, will rebase before submitting.
+[[nodiscard]] static uint64_t MonotonicTimestampNs() { return 0; }
+
+namespace orbit::tracing {
+Listener::Listener(const TimerCallback& callback) {
+  // Note: only one listener is supported.
+  absl::MutexLock lock(&mutex_);
+  CHECK(g_callback == nullptr);
+  callback_ = callback;
+  g_callback = &callback_;
+}
+
+Listener::~Listener() {
+  absl::MutexLock lock(&mutex_);
+  g_callback = nullptr;
+}
+
+}  // namespace orbit::tracing
+
+namespace orbit_api {
+
+void Start(const char* name, uint64_t, orbit::Color color) {
+  scopes.emplace_back(orbit::tracing::Scope());
+  auto& scope = scopes.back();
+  scope.name = name;
+  scope.begin = MonotonicTimestampNs();
+  scope.color = color;
+}
+
+void Stop() {
+  auto& scope = scopes.back();
+  scope.end = MonotonicTimestampNs();
+  scope.depth = scopes.size() - 1;
+  scope.tid = absl::base_internal::GetTID();  // Will be replaced by our own implementation (#1113).
+  scope.type = ScopeType::kScope;
+
+  absl::Mutex* mutex = orbit::tracing::Listener::GetMutex();
+  mutex->Lock();
+  if (g_callback != nullptr) (*g_callback)(std::move(scope));
+  mutex->Unlock();
+
+  scopes.pop_back();
+}
+
+// Will be implemented shortly.
+void StartAsync(const char*, uint64_t, orbit::Color) { CHECK(0); }
+void StopAsync(uint64_t) { CHECK(0); }
+
+void TrackScope(const char* name, uint64_t value, orbit::Color color, ScopeType type) {
+  orbit::tracing::Scope scope;
+  scope.begin = MonotonicTimestampNs();
+  scope.name = name;
+  scope.color = color;
+  scope.tracked_value = value;
+  scope.tid = absl::base_internal::GetTID();
+  scope.type = type;
+
+  absl::MutexLock lock(orbit::tracing::Listener::GetMutex());
```

Comment:
this can be function-local static mutex if this is the only place where we guard - if not, the functionality should probably be incapacitated into class owning this mutex.